typing/ikinds/product_lattice.ml

```diff
@@ -0,0 +1,80 @@
+module type SHAPE = sig
+  val axis_sizes : int array (* n_i >= 1 *)
+end
+
+(* Simple array-based product lattice for debugging/prototyping. *)
+module Make (S : SHAPE) = struct
+  type t = int array
+
+  let axis_sizes = Array.copy S.axis_sizes
```

Comment:
Maybe use immutable arrays to avoid the need to copy? I think these sizes should never change.